packages/editor/src/hooks/custom-sources-backwards-compatibility.js

```diff
@@ -1,76 +1,95 @@
+/**
+ * External dependencies
+ */
+import { pickBy, mapValues, isEmpty, mapKeys } from 'lodash';
+
 /**
  * WordPress dependencies
  */
-import { getBlockType } from '@wordpress/blocks';
 import { useSelect } from '@wordpress/data';
 import { useEntityProp } from '@wordpress/core-data';
-import { useMemo, useCallback } from '@wordpress/element';
+import { useMemo } from '@wordpress/element';
 import { createHigherOrderComponent } from '@wordpress/compose';
 import { addFilter } from '@wordpress/hooks';
 
-const EMPTY_OBJECT = {};
-function useMetaAttributeSource( name, _attributes, _setAttributes ) {
-	const { attributes: attributeTypes = EMPTY_OBJECT } =
-		getBlockType( name ) || EMPTY_OBJECT;
-	let [ attributes, setAttributes ] = [ _attributes, _setAttributes ];
+/** @typedef {import('@wordpress/compose').WPHigherOrderComponent} WPHigherOrderComponent */
+
+/** @typedef {import('@wordpress/blocks').WPBlockSettings} WPBlockSettings */
+
+/**
+ * Object mapping an attribute key to its corresponding meta property.
+ *
+ * @typedef {Object<string,string>} WPMetaAttributeMapping
+ */
 
-	if ( Object.values( attributeTypes ).some( ( type ) => type.source === 'meta' ) ) {
-		// eslint-disable-next-line react-hooks/rules-of-hooks
-		const type = useSelect( ( select ) => select( 'core/editor' ).getCurrentPostType(), [] );
-		// eslint-disable-next-line react-hooks/rules-of-hooks
-		const [ meta, setMeta ] = useEntityProp( 'postType', type, 'meta' );
+/**
+ * Given a meta attribute mapping, returns a new higher-order component which
+ * manages attributes merging and updates to consume from and mirror to the
+ * associated entity record.
+ *
+ * @param {WPMetaAttributeMapping} metaKeys Meta attribute mapping.
+ *
+ * @return {WPHigherOrderComponent} Higher-order component.
+ */
+const createWithMetaAttributeSource = ( metaKeys ) => createHigherOrderComponent(
+	( BlockEdit ) => ( { attributes, setAttributes, ...props } ) => {
+		const postType = useSelect( ( select ) => select( 'core/editor' ).getCurrentPostType() );
```

Comment:
Hm, do you suggest there's a specific need here, or that we should always tend to default to passing an empty array as the second argument? Looking through to `useSelect` to see how what impact this has ([source](https://github.com/WordPress/gutenberg/blob/68d7632dcc598799df80801c838747b94fef5264/packages/data/src/components/use-select/index.js#L78)), I guess the potential concern here is that we'd otherwise be creating a fresh callback reference to `mapSelect` on every render? I guess this arguments usage is pretty standard fare with hooks, but I kinda see it as a non-obvious awkwardness of their use.

Not a sword I care to fall on in the context of this pull request, though! 😄 I'll make the suggested edit.

Reply:
Without it, we run `mapSelect` on every render.

Reply:
> Without it, we run `mapSelect` on every render.

That's a pretty concerning default behavior for `useSelect`.

Reply:
That's how all hooks `deps` arrays work. So that your callback can have closures by default.

Reply:
> That's how all hooks `deps` arrays work. So that your callback can have closures by default.

Yeah. Still awkward. I guess I just need to force it into my muscle memory.

Reply:
We can always add the `eslint` rule: `react-hooks/exhaustive-deps`.

Reply:
> We can always add the `eslint` rule: `react-hooks/exhaustive-deps`.

Yeah, I'm a bit curious about that. It was included previously, then later removed at https://github.com/WordPress/gutenberg/pull/14985#discussion_r355242262 . Seems like something we should want?

Reply:
Yeah